### ISSUE TITLE
PaddleInferTensor能不能提供一个copy_from_tensor这样的功能

### ISSUE THREAD

**Issue by wuqi930907** [PaddlePaddle/PaddleSeg#1901]
目前使用numpy在cpu上做transforms非常耗时，我想把这些transforms放到GPU上去实现，那PaddleInferTensor能不能提供一个类似于copy_from_tensor的函数，用来接收GPU上的paddle.tensor类型数据，或者接收Cupy类型的数据呢

**Comment by shiyutang** [comment]
If you want to use GPU in transform, you can try "import cupy as np" in the transform file. And I am not sure why do you need copy_from_tensor?

**Comment by wuqi930907** [comment]
@shiyutang 谢谢您的回答，因为我发现用cupy或者paddle.tensor(gpu版)进行预处理之后，再从cupy或者paddle.tensor转为numpy非常的耗时，而PaddleInferTensor应该也是将输入放入gpu这样的一个功能（应该是吧），所以为何预处理完成后还要将数据从gpu->cpu->gpu呢，直接PaddleInferTensor拷贝gpu内存会更有效率吧

**Comment by shiyutang** [comment]
请问你说的预处理是哪一部分的代码呢，我针对性查看一下

**Comment by wuqi930907** [comment]
@shiyutang  您好，推理一张图片通常分三个步骤，预处理->推理->后处理。因为我看目前代码预处理是放在cpu上用numpy去处理的，速度很慢，所以我希望把预处理放到GPU上处理，从而更好地利用GPU，所以我进行了以下改动：
动态图推理改动后步骤：
1.将numpy图片转为paddle.tensor(CUDA)，在GPU上做Normalize；
2.用paddle.tensor(CUDA)进行推理。（测试的paddleseg/core/predict.py中logits = model(im)此行代码速度）
3.后处理。
动态图推理问题现象：照理来讲，图片的大小会影响推理那一步的速度，但是目前在我的机器上测试，一张很大的图片和一张很小的图片推理速度都是40ms，是不是动态图推理有什么异步机制。速度固定代表我推理很小的图片时FPS会很低。
静态图推理改动后步骤：
1.将numpy图片转为paddle.tensor(CUDA)，在GPU上做Normalize；
2.将paddle.tensor(CUDA)转为numpy；
3.将numpy数据进行推理；
4.后处理。
静态图推理问题：第2步中将paddle.tensor(CUDA)转为numpy消耗了时间，但我觉得这个时间消耗是不必要的，因为PaddleInferPredictor推理时也需要把numpy数据放入GPU中。

所以我有以下的问题：
1.因为我是刚接触paddlepaddle，所以请问下paddle的动态图推理是不是确实有我上面说的现象，推理速度固定，因为torch并不会有这种现象。
2.静态图推理能不能接受paddle.tensor(CUDA)类型的输入数据。

谢谢

**Comment by juncaipeng** [comment]
动态图推理，不同输入大小，速度应该不一致的。
静态图推理中，inference不支持直接使用gpu上的数据作为输入，这样会导致内存无法管理。
既然你已经使用Paddle API实现了对GPU Tensor进行Normalize，建议将该部分代码放到组网中，这样导出模型时直接有Normalize，部署是不再需要预处理过程了。

**Comment by wuqi930907** [comment]
@juncaipeng 谢谢您的建议，目前我是把Normalize放入网络中了，但是我换了几台机器测试paddleseg/core/predict.py中logits = model(im)这行的速度，仍然得不到我想要的结果，实在是有些想不通，也许是因为我是paddle的新手吧，我会继续尝试验证的，谢谢

**Comment by wuqi930907** [comment]
还是试验了动态图推理，想不明白，求大神解答 @juncaipeng @shiyutang
代码如下：
import cv2
import numpy as np
import paddle
import time
import paddle.nn.functional as F

from paddleseg.models import HarDNet

im_path = '35.jpg'
model_path = 'output_new/train_model/hardnet/best_model/model.pdparams'
crop_size = (512, 1024)
mean = (0.5, 0.5, 0.5)
std = (0.5, 0.5, 0.5)
num_class = 20

# 加载模型
model = HarDNet(num_class)
para_state_dict = paddle.load(model_path)
model.set_dict(para_state_dict)
model.eval()

# 开始推理
with paddle.no_grad():
    # 读取图像
    im = cv2.imread(im_path)
    ori_shape = im.shape[:2]

    # 图像预处理
    im = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)
    im = cv2.resize(im, crop_size, cv2.INTER_LINEAR)
    im = im.astype(np.float32, copy=False) / 255.0
    im -= mean
    im /= std

    # 转换成4通道张量格式
    im = np.transpose(im, (2, 0, 1))
    im = im[np.newaxis, ...]
    im = paddle.to_tensor(im)

    # 模型推理
    for i in range(20):
        st_time = time.time()
        logits = model(im)
        logit = logits[0]
        print("cost:{}".format((time.time() - st_time) * 1000))

        pred = paddle.argmax(logit, axis=1, keepdim=True, dtype='int32')
        pred = F.interpolate(pred, ori_shape, mode='nearest')

        # 二值掩码图像保存
        pred = paddle.squeeze(pred)
        pred = pred.numpy().astype('uint8') * 255

    cv2.imwrite('output/result/result.jpg', pred)

问题是当我无论怎么修改crop_size 的大小时，打印出来的cost永远是不变的。
当crop_size = (512, 1024)时log如下：
![1648625318(1)](https://user-images.githubusercontent.com/34122261/160775818-d0303e0e-03af-4a30-9a14-565e27a44deb.png)

当crop_size = (214, 214)时log如下：
![1648625289(1)](https://user-images.githubusercontent.com/34122261/160775747-602b1e56-c406-4d62-99ba-6b3ed071c079.png)


两种图片大小推理那行代码执行舒服无太大区别，而用PaddleInferPredictor推理时，两种图片大小的推理速度差异是很大的

**Comment by juncaipeng** [comment]
前者是动态图执行前向，后者是静态图执行推理。
通常大家不会关注前者的速度，都是使用后者方式进行部署。

**Comment by wuqi930907** [comment]
@juncaipeng 好的，十分感谢，不过这个机制好像和torch不太一样，写法差不多，但是torch前向执行的时候符合FPS与计算量成反比，paddle好像不满足这个规律

**Comment by Zhang-O** [comment]
```python
paddle.set_device("cpu")

x = np.ones([100,100], dtype=np.float32)
paddle.device.cuda.synchronize()
t0 = time.time()
xx = paddle.to_tensor(preds)
paddle.device.cuda.synchronize()
t1 = time.time()
print(f" t1 - t0 take time: { t1 - t0 }")
```
paddle.to_tensor 这个操作非常耗时 50ms左右，感觉比torch.from_numpy 慢很多
